File: 1329-sort-the-matrix-diagonally/1329-sort-the-matrix-diagonally.cpp
#include <algorithm>
class Solution {
public:
    void arraySort(int x, int y, int size, vector<vector<int>>& mat, vector<int>& temp) {
        for (int i = 0; i < size; i++) {
            temp[i] = (mat[y + i][x + i]);
        }

        sort(temp.begin(), temp.begin() + size);

        for (int i = 0; i < size; i++) {
            mat[y + i][x + i] = temp[i];
        }
    }
    
    vector<vector<int>> diagonalSort(vector<vector<int>>& mat) {
        int xSize = mat[0].size();
        int ySize = mat.size();
        int size = 0;
        
        if (xSize > ySize) {
            size = ySize;
        } else {
            size = xSize;
        }

        vector<int> tempArray(size);
        if (xSize == 1 || ySize == 1) {
            return mat;
        }
        if (xSize > ySize) {
            arraySort(0, 0, ySize, mat, tempArray);
        } else {
            arraySort(0, 0, xSize, mat, tempArray);
        }

        for (int i = 1; i < xSize; i++) {
            arraySort(i, 0, (xSize - i) < ySize ? (xSize - i) : ySize, mat, tempArray);
        }
        for (int i = 1; i < ySize; i++) {
            arraySort(0, i, (ySize - i) < xSize ? (ySize - i) : xSize, mat, tempArray);
        }
        return mat;
    }
};

File: 1329-sort-the-matrix-diagonally/1329-sort-the-matrix-diagonally.java
import java.util.Arrays;

class Solution {
    public void arraySort(int x, int y, int size, int[][] mat) {
        int[] tempArray = new int[size];
        for (int i = 0; i < size; i++) {
            tempArray[i] = mat[y + i][x + i];
        }

        Arrays.sort(tempArray);

        for (int i = 0; i < size; i++) {
            mat[y + i][x + i] = tempArray[i];
        }
    }
    public int[][] diagonalSort(int[][] mat) {
        int xSize = mat[0].length;
        int ySize = mat.length;

        if (xSize == 1 || ySize == 1) {
            return mat;
        }
        if (xSize > ySize) {
            arraySort(0, 0, ySize, mat);
        } else {
            arraySort(0, 0, xSize, mat);
        }

        for (int i = 1; i < xSize; i++) {
            arraySort(i, 0, (xSize - i) < ySize ? (xSize - i) : ySize, mat);
        }
        for (int i = 1; i < ySize; i++) {
            arraySort(0, i, (ySize - i) < xSize ? (ySize - i) : xSize, mat);
        }
        return mat;
    }
}